core-rust/state-manager/src/jni/state_reader.rs

```diff
@@ -0,0 +1,110 @@
+ * shall be liable for any direct or indirect, special, incidental, consequential or other
+ * losses of any kind, in tort, contract or otherwise (including but not limited to loss
+ * of revenue, income or profits, or loss of use or data, or loss of reputation, or loss
+ * of any economic or other opportunity of whatsoever nature or howsoever arising), arising
+ * out of or in connection with (without limitation of any use, misuse, of any ledger system
+ * or use made or its functionality or any performance or operation of any code or protocol
+ * caused by bugs or programming or logic errors or otherwise);
+ *
+ * A. any offer, purchase, holding, use, sale, exchange or transmission of any
+ * cryptographic keys, tokens or assets created, exchanged, stored or arising from any
+ * interaction with the Work;
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+use jni::objects::{JClass, JObject};
+use jni::sys::jbyteArray;
+use jni::JNIEnv;
+use radix_engine::system::system_substates::FieldSubstate;
+use radix_engine::types::*;
+
+use radix_engine_store_interface::db_key_mapper::*;
+
+use radix_engine::blueprints::consensus_manager::{
+    ValidatorField, ValidatorProtocolUpdateReadinessSignalFieldPayload,
+};
+
+use node_common::java::*;
+
+use super::node_rust_environment::JNINodeRustEnvironment;
+
+//
+// JNI Interface
+//
+
+#[no_mangle]
+extern "system" fn Java_com_radixdlt_state_RustStateReader_getValidatorProtocolUpdateReadinessSignal(
+    env: JNIEnv,
+    _class: JClass,
+    j_node_rust_env: JObject,
+    request_payload: jbyteArray,
+) -> jbyteArray {
+    jni_sbor_coded_fallible_call(
+        &env,
+        request_payload,
+        |validator_address: ComponentAddress| -> JavaResult<Option<String>> {
+            let result = JNINodeRustEnvironment::get(&env, j_node_rust_env)
+                .state_manager
+                .database
+                .read_current()
```

Comment:
We could consider using `access_non_locked_historical` to avoid locking if we were worried about locking on the health endpoint.